engine/src/main/java/graphql/nadel/engine/execution/NadelExecutionStrategy.java

```diff
@@ -129,37 +140,89 @@ public NadelExecutionStrategy(List<Service> services,
-                ServiceOrError serviceOrError = assertNotNull(serviceExecutionHooks.resolveServiceForField(services, fieldExecutionStepInfo));
+        }
+        return Async.each(result);
+    }
 
-                if(serviceOrError.getService() != null) {
-                    service = serviceOrError.getService();
-                } else {
-                    GraphQLError graphQLError = assertNotNull(serviceOrError.getError());
+    private CompletableFuture<OneServiceExecution> prepareDynamicServiceExecution(ExecutionContext executionCtx, ExecutionStepInfo rootExecutionStepInfo, ExecutionStepInfo fieldExecutionStepInfo) {
+        ServiceOrError serviceOrError = assertNotNull(
+                serviceExecutionHooks.resolveServiceForField(services, fieldExecutionStepInfo),
+                () -> "Service resolution hook must never return null."
+        );
 
-                    result.add(CompletableFuture.completedFuture(new OneServiceExecution(null, null, null, true, graphQLError, fieldExecutionStepInfo)));
-                    continue;
-                }
+        if (serviceOrError.getService() == null) {
+            GraphQLError graphQLError = assertNotNull(serviceOrError.getError(), () -> "Hook must return an error object when Service is null");
 
-            } else {
-                service = getServiceForFieldDefinition(fieldExecutionStepInfo.getFieldDefinition());
-            }
+            return CompletableFuture.completedFuture(new OneServiceExecution(null, null, null, true, graphQLError, fieldExecutionStepInfo));
+        }
 
-            CreateServiceContextParams parameters = CreateServiceContextParams.newParameters()
-                    .from(executionCtx)
-                    .service(service)
-                    .executionStepInfo(fieldExecutionStepInfo)
-                    .build();
+        Assert.assertTrue(
+                fieldExecutionStepInfo.getUnwrappedNonNullType() instanceof GraphQLInterfaceType,
+                () -> format("field annotated with %s directive is expected to be of GraphQLInterfaceType", DYNAMIC_SERVICE_DIRECTIVE_DEFINITION.getName())
+        );
 
-            CompletableFuture<Object> serviceContextCF = serviceExecutionHooks.createServiceContext(parameters);
-            CompletableFuture<OneServiceExecution> serviceCF = serviceContextCF.thenApply(serviceContext -> new OneServiceExecution(service, serviceContext, fieldExecutionStepInfo, false, null, fieldExecutionStepInfo));
-            result.add(serviceCF);
-        }
-        return Async.each(result);
+        Service service = serviceOrError.getService();
+
+        List<String> serviceObjectTypes = service.getDefinitionRegistry().getDefinitions()
+                .stream()
+                .filter(definition -> definition instanceof ObjectTypeDefinition)
+                .map(definition -> (ObjectTypeDefinition) definition)
+                .map(ObjectTypeDefinition::getName)
+                .collect(toList());
```

Comment:
@temaEmelyan had a new helper method here `getDefintions(ObjectTypeDefinition.class)` which could simplify this.

Its ok to leave this in - be nice to use the new one however later

Reply:
Thanks for the heads up! I've copied that method into `DefinitionRegistry`. 
It should not result in merge conflicts doesn't matter which PR gets merged first 🤞